nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+        .map_err(ActionError::action_failed)?;
+
+    warn!(log, "disk volume construction request {}", disk_volume.data());
+
+    let disk_volume_construction_request =
+        serde_json::from_str(&disk_volume.data()).map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&format!(
+                "failed to deserialize disk {} volume data: {}",
+                disk.id(),
+                e,
+            )))
+        })?;
+
+    // The volume construction request must then be modified to point to the
+    // read-only crucible agent downstairs (corresponding to this snapshot)
+    // launched through this saga.
+    let replace_sockets_map =
+        sagactx.lookup::<BTreeMap<String, String>>("replace_sockets_map")?;
+    let volume_construction_request: VolumeConstructionRequest =
+        replace_sockets(
+            &disk_volume_construction_request,
+            &replace_sockets_map,
+        )
+        .map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&e.to_string()))
+        })?;
+
+    // Create the volume record for this snapshot
+    let volume_data: String =
+        serde_json::to_string(&volume_construction_request).map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&e.to_string()))
+        })?;
+    warn!(log, "snapshot volume construction request {}", volume_data);
+    let volume = db::model::Volume::new(volume_id, volume_data);
+
+    let volume_created = osagactx
+        .datastore()
+        .volume_create(volume)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    warn!(log, "volume {} created ok", volume_id);
+
+    Ok(volume_created)
+}
+
+async fn ssc_create_volume_record_undo(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let volume_id = sagactx.lookup::<Uuid>("volume_id")?;
+    warn!(log, "deleting volume {}", volume_id);
+    osagactx.datastore().volume_delete(volume_id).await?;
+
+    Ok(())
+}
+
+async fn ssc_finalize_snapshot_record(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<db::model::Snapshot, ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    warn!(log, "snapshot final lookup...");
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+    let (.., authz_snapshot) = LookupPath::new(&opctx, &osagactx.datastore())
+        .snapshot_id(snapshot_id)
+        .lookup_for(authz::Action::Modify)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    warn!(log, "snapshot final lookup ok");
+
+    let snapshot = osagactx
+        .datastore()
+        .project_snapshot_update_state(
+            &opctx,
+            &authz_snapshot,
+            db::model::Generation::new(),
```

Comment:
By demanding that the old generation number be an exact value, I think this saga node is also not idempotent. Executing this node twice would result in the second invocation seeing "not found", and returning an error.

Reply:
addressed in 9962ad8e